org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java

```diff
@@ -0,0 +1,145 @@
+package org.metaborg.spoofax.shell.commands;
+
+import java.io.IOException;
+import java.util.Map;
+
+import org.metaborg.core.MetaborgException;
+import org.metaborg.core.context.IContext;
+import org.metaborg.core.context.IContextService;
+import org.metaborg.core.language.ILanguageImpl;
+import org.metaborg.core.project.IProject;
+import org.metaborg.spoofax.core.shell.ShellFacet;
+import org.metaborg.spoofax.core.unit.ISpoofaxAnalyzeUnit;
+import org.metaborg.spoofax.core.unit.ISpoofaxParseUnit;
+import org.metaborg.spoofax.core.unit.UnitWrapper;
+import org.metaborg.spoofax.shell.core.EvaluateUnit;
+import org.metaborg.spoofax.shell.core.IEvaluationStrategy;
+import org.metaborg.spoofax.shell.hooks.IResultHook;
+import org.metaborg.spoofax.shell.invoker.ICommandFactory;
+import org.metaborg.spoofax.shell.output.AnalyzeResult;
+import org.metaborg.spoofax.shell.output.EvaluateResult;
+import org.metaborg.spoofax.shell.output.IResultFactory;
+import org.metaborg.spoofax.shell.output.InputResult;
+import org.metaborg.spoofax.shell.output.ParseResult;
+import org.spoofax.interpreter.terms.IStrategoTerm;
+
+import com.google.inject.Inject;
+import com.google.inject.assistedinject.Assisted;
+
+/**
+ * Represents an evaluate command sent to Spoofax.
+ */
+public class EvaluateCommand extends SpoofaxCommand {
+    private static final String DESCRIPTION = "Evaluate an expression";
+
+    private IContextService contextService;
+    private ParseCommand parseCommand;
+    private AnalyzeCommand analyzeCommand;
+    private IResultFactory unitFactory;
+
+    @Inject
+    private Map<String, IEvaluationStrategy> evaluationStrategies;
+    private EvaluationInvocationStrategy invocationStrategy;
+
+    /**
+     * Interface for what happens before evaluation of parsed input (i.e. either analyze or evaluate
+     * as-is).
+     */
+    private interface EvaluationInvocationStrategy {
+        EvaluateResult performEvaluation(IContext context, ParseResult parsed,
+                                         IEvaluationStrategy evalStrategy)
+            throws MetaborgException;
+    }
+
+    /**
+     * Analyze before invoking the evaluation strategy.
+     */
+    private class AnalyzedInvocationStrategy implements EvaluationInvocationStrategy {
+        @Override
+        public EvaluateResult performEvaluation(IContext context, ParseResult parsed,
+                                                IEvaluationStrategy evalStrategy)
+            throws MetaborgException {
+            AnalyzeResult analyzed = analyzeCommand.analyze(parsed);
+            IStrategoTerm ast = evalStrategy.evaluate(analyzed, context);
+            // TODO: Normally this is done in the unit service, but since this is my own Unit this
```

Comment:
So this would require a change in Spoofax itself? If so, is this doable for us, or does it work good enough as-is / does Spoofax itself have no use for it besides the shell?